### DISCUSSION TITLE
How to use Remix Link components in Storybook

### DISCUSSION THREAD

**Discussion by tachibanayu24** [remix-run/remix#6273] [Q&A] [answered]
Hello.

I am building UIs using the Link component provided by `remix-run/react` and would like to add UI components to Storybook.
Just adding Link.stories.tsx throws the error `useHref() may only be used in the context of a <Router> component.` 
I would like to know how to avoid this error.

What I have tried is described below.

Just adding Link.stories.tsx will result in throwing an error `useHref() may only be used in the context of a <Router> component.` 

For the Link component of `react-router`, a Router component is provided, which can be wrapped in `.storybook/preview` or *.stories.tsx, but I am not sure if there is such a Provider in `@remix-run/react`.

So I tried to use `@remix-run/testing` `unstable_createRemixStub`, which seems to have been added recently.

```tsx
export const decorators: Decorator[] = [
  (Story: StoryFn) => {
    const route = [
      {
        path: '/*', element: <Story />, element: <Story />, element: <StoryFn
        element: <Story />, }
      }, }
    ]
    const RemixStub = unstable_createRemixStub(route)
    return <RemixStub initialEntries={['/']} />
  }, }
]
```

This way, something similar to Remix's initial error text is now displayed on the Storybook, but as for the useHref error text that is thrown, it is unchanged from the first time.

![Screenshot 2023-05-02 at 14 40 41](https://user-images.githubusercontent.com/23476543/235589002-e71e3398-1e09-425e-a707-5fa35586c6fd.png)

Do you have any ideas?

Thanks.

**Comment by DotwoodMedia** [comment]
Add a decorator to your storybook preview file
```
import type { Preview } from "@storybook/react";
import { MemoryRouter } from "react-router";
import React from "react";

const preview: Preview = {
  parameters: {
    actions: { argTypesRegex: "^on[A-Z].*" },
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/,
      },
    },
  },
  decorators: [
    (story) => <MemoryRouter initialEntries={["/"]}>{story()}</MemoryRouter>,
  ],
};

export default preview;
```

**Reply by tachibanayu24** [reply]
@DotwoodMedia 
Thank you for your comment.

The component I am trying to add to the storybook uses the Link component from `@remix-run/react`, so wrapping it in a Router of react-router does not seem to resolve the error of not being able to reference useHref It is.

I was wondering if using RemixStub would solve the problem, but that was not possible with the same error as described above.

**Comment by tachibanayu24** [comment]
I think there is room to argument that UI components such as those added to storybook should not depend on a framework like remix, but should use react-router.
But be that as it may, is there any way to use remix's Link in storybook?

**Comment by florianchevallier** [comment]
I don't know if it will helps in any way but on my side I did it this way : 

```tsx
// .storybook/preview.tsx
import { unstable_createRemixStub } from '@remix-run/testing';
import { Preview } from '@storybook/react';
import '../app/tailwind.css';

const preview: Preview = {
  decorators: [
    Story => {
      const RemixStub = unstable_createRemixStub([
        {
          path: '/*',
          element: <Story />,
        },
      ]);

      return <div><RemixStub initialEntries={ ['/']} /></div>;
    },
  ],
};

export default preview;
```


And then 
```tsx
// Button.stories.tsx

import type { Meta, StoryObj } from '@storybook/react';
import Button from '../../../app/component/Button';

export default {
  title: 'Composants/Button/Primary',
  component: Button,
} as Meta;

type Story = StoryObj<typeof Button>;

export const Plain: Story = {
  args: {
    label: 'Button',
  },
};

export const Loading: Story = {
  args: {
    ...Plain.args,
    loading: true,
  },
};

👇 this one used to have the error as it's a <Link />
export const Link: Story = {
  args: {
    ...Plain.args,
    to: 'https://www.google.fr',
  },
};
```

hope it will help !

**Reply by tachibanayu24** [reply]
@florianchevallier 
Thank you for your response!

Your code and mine look almost same, but I copied your code and tried it.
As a result, it did not work in my environment and I kept getting the same error. 😭 
I thought it might be an older version of Remix, but changing to 1.16.0 did not change the result.

My link component just wraps the `@remix-run/react` link almost , so I don't know why the error occurs.
Hmm...

**Reply by simonepizzamiglio** [reply, marked as answer]
I'm on Remix 1.16.0 and Storybook 7.0.11, the following code works for me.

```ts
// Link.stories.tsx

import type { StoryObj, Meta } from '@storybook/react';
import { unstable_createRemixStub as createRemixStub } from '@remix-run/testing';
import { Link } from './Link';

const meta: Meta<typeof Link> = {
  title: 'components/Link',
  component: Link,
  decorators: [
    (Story) => {
      const RemixStub = createRemixStub([
        {
          path: '/',
          element: <Story />,
        },
      ]);

      return <RemixStub />;
    },
  ],
  tags: ['autodocs'],
  args: {
    children: "I'm a link",
    to: '/',
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {};
```

**Reply by tachibanayu24** [reply]
Thank you for your response!

I updated the storybook version to the latest (7.0.12) and it works as expected.
I should have tried this earlier 🙇

**Reply by maxxborer** [reply]
How to work with this now? In version 2 and higher there is no longer an element property.
<img width="1169" alt="изображение" src="https://github.com/remix-run/remix/assets/44510263/8ea66e12-e9f0-45c6-8da6-d32b25a600bc">
<img width="1077" alt="изображение" src="https://github.com/remix-run/remix/assets/44510263/3115d5ac-8639-4744-9f15-f836513b1a8f">

**Reply by maxxborer** [reply]
Found the solution in the source code:
```js
const preview: Preview = {
  decorators: [
    (Story) => {
      const RemixStub = createRemixStub([
        {
          path: '/*',
          action: () => ({ redirect: '/' }),
          loader: () => ({ redirect: '/' }),
          Component() { // <-- here
            return <Story />;
          }
        },
      ]);

      return (
        <Providers>
          <RemixStub />
        </Providers>
      )
    },
  ],
};

export default preview;
```

**Reply by RyanofWoods** [reply]
Where did you see this in the source code @maxxborer? Wondering about the `Providers` usage.

**Reply by maxxborer** [reply]
@RyanofWoods 

I found something like this and instead of specifying a link to the component, I made it:
```js
Component: MyComponent,
```
as a method:
```jsx
Component() { // <-- here
  return <Story />;
},
```

From this file:
https://github.com/remix-run/remix/blob/49d638b8d02a7b0b34d8da797e1f676d5e1fff71/docs/other-api/testing.md?plain=1#L18

Code from file for example:
```js
const RemixStub = createRemixStub([
  {
    path: "/",
    Component: MyComponent,
    loader() {
      return json({ message: "hello" });
    },
  },
]);
```

**Reply by RyanofWoods** [reply]
Thanks for the reply @maxxborer. Unfortunately, this still doesn't work for me 🙁  Storybook `7.0.12`. I have tried various setups with @remix-run/testing versions `2.2.x`, `2.0.1` and `1.19.1` 🤔 I still keep getting
`Error: useLocation() may be used only in the context of a <Router> component.`

**Reply by maxxborer** [reply]
@RyanofWoods 
you can try to see this implementation in my project, maybe there is something interesting:
https://github.com/gorakuicu/gorakuicu/blob/main/apps/web/.storybook/preview.tsx

**Reply by RyanofWoods** [reply]
Thanks for your help @maxxborer, I appreciate it. I managed to solve it by using `v1.19.1`. Its dependencies were installed incorrectly and gave me issues. After deleting them and installing them again, it worked 👍

**Reply by lukeclifton** [reply]
Hey guys. we're using Storybook version `7.6.4` and `@remix-run/testing` version `2.3.1` and this doesn't work for us. We still see the same error.

Any ideas?

**Reply by lukeclifton** [reply]
Ok we found our issue. Updating `@remix-run/react` from version `2.1.0` to version `2.3.1` fixed this error for us.

**Comment by coindegen** [comment]
I attempted the above solutions but was getting all kinds of Typescript errors like this:

```
'RemixStub' refers to a value, but is being used as a type here. Did you mean 'typeof RemixStub'?ts(2749)

Type '() => StoryAnnotations<ReactRenderer, { ... }>' is not assignable to type 'ComponentType<{}> | null | undefined'.
```

Not saying this is correct, but this is the only way I could satisfy the transpiler:

```ts
import type { Meta, StoryObj } from "@storybook/react";
import { CollectionCard } from "../app/theme/components/CollectionCard/CollectionCard";
import { MOCK_COLLECTION } from "./mocks/mocks";
import { unstable_createRemixStub as createRemixStub } from "@remix-run/testing";

const meta = {
  title: "Example/CollectionCard",
  component: CollectionCard,
  decorators: [
    (story) => {
      const remixStub = createRemixStub([
        {
          path: "/*",
          action: () => ({ redirect: "/" }),
          loader: () => ({ redirect: "/" }),
          Component: () => story(),
        },
      ]);

      return remixStub({ initialEntries: ["/"] });
    },
  ],
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  tags: ["autodocs"],
} satisfies Meta<typeof CollectionCard>;
```

**Reply by AYastrebtsov** [reply]
Just started a new project, works like a charm!

**Reply by UnworldlyEel** [reply]
thank you DEGEN, you saved me a bunch of frustration!

**Comment by lorenzo-del-rosario** [comment]
I was having the same issue with `"@remix-run/testing": "^2.4.1"`, even after using the new `Component` argument that came out with version 2.
Downgrading ` "@remix-run/testing": "^2.1.0"` with `"@remix-run/react": "2.1.0"` solved the issue.

**Comment by thegrid22593** [comment]
If you are using storybook 8, you need to migrate to using [this package from storybook](https://storybook.js.org/addons/storybook-addon-remix-react-router).

Also, be sure to be on the same version of any package prefixed with @remix-run/

Here is an example:

```
import { withRouter } from 'storybook-addon-remix-react-router';

const meta: Meta<typeof TextLink> = {
  component: TextLink,
  title: 'TextLink',
  tags: ['test'],
  decorators: [withRouter]
};
```

Also, if you are adding withRouter to the decorator `preview.ts` it will not work with any `index.spec.tsx` tests. You will have to apply the `withRouter` decorator to the component meta itself.

**Reply by nickfujita** [reply]
Thanks for this suggestion. I just tried it out, but for some reason it's not working with Remix, only react-router.
More details: https://github.com/JesusTheHun/storybook-addon-remix-react-router/issues/74

Curious if there was any specific gotchas to look out for? What did you mean by `Also, be sure to be on the same version of any package prefixed with @remix-run/`?

**Reply by thegrid22593** [reply]
@nickfujita - what I meant is when using remix there are a few packages installed. In our case some include

`@remix-run/react`
`@remix-run/serve`
`@remix-run/node` 
`@remix-run/testing`

All of these and any others you may have should all be on the same version, in our case its `2.13.1` which is latest at the time of this comment, but should work with minor versions off V2.

**Reply by alejoFernandezSS** [reply]
I'm trying to make this work on Remix v2.15.1 with Storybook v8.2.9 but I'm getting the following error and I have no clue why or where to start debugging

`✘ [ERROR] No matching export in "../../node_modules/.pnpm/react-router@6.28.0_react@18.3.1/node_modules/react-router/dist/index.js" for import "useSearchParams"`

Previous to using the storybook addon I was getting the typical error of `usehref() may be used only in the context of a <router> component`
